### PR TITLE
Integration for Bedrock Authentication

### PR DESCRIPTION
**Pull request by Andrewngabriel**
Apologies in advance for the massive PR 🙂

Here are the main high level items I worked on:
- Added support for authentication with Bedrock API.
  - ✅ Login
  - ✅ Signup
  - ✅ Password Reset
- Added shops and products screens showing data directly from Bedrock API using JWT token stored on device using [flutter_secure_storage](https://pub.dev/packages/flutter_secure_storage) which uses Keychain for iOS and Keystore on Android.
- Added unit tests for Authentication controller by mocking real API behavior.
- Added Github Action workflow for running unit & widget tests on pull requests.
- Updated readme to include feature highlights and instructions for getting started.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,11 +50,14 @@ samples, guidance on mobile development, and a full API reference.
   ```
   `--dart-define` is used to pass environment configuration.
 
-#### Environment variables
+#### ⚙️ Environment variables
 
-- `BEDROCK_API`: This should be pointed to the base url of where the API is running.
+| Variable              | Default Value          | Description |
+| -                     | -                      | -           |
+| `BEDROCK_API`         | https://localhost:2300 | Base url of where the API is running |
```
Comment (dominiek) [inline README.md:64]: I think this is good, but I think we should have some sort of facility inside the base app to switch between development, staging and production. It is quite common that we give out the app to early users where we want them to explicitly use one of those environments. 

We can do this in a different PR set

**Review by dominiek — APPROVED** [review]
This looks good to me, but I may not be the best person to code review this

File: lib/src/auth/auth_controller.dart [hunk 2]
```diff
@@ -1,34 +1,141 @@
...
-    _authenticated = !_authenticated;
+  @override
+  Future<void> register(
+    String email,
+    String firstName,
+    String lastName,
+    String password,
+  ) async {
+    http.Response? response = await apiService.post('/auth/register', payload: {
+      'email': email,
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_controller.dart]: You think this could be a model class class called for example `registration_request.dart`  and maybe the `email` `firstName` and `lastName` could be extracted to another model called `User`? Im saying this because this way, the model could be reused in other places, specially the `User`.

File: lib/src/auth/auth_controller.dart [hunk 3]
```diff
@@ -1,34 +1,141 @@
...
-  login(String email, String password) {}
-  register(String email, String name, String password) {}
-  logout() {}
-  resetPassword(String email) {}
-  setPassword(String newPassword) {}
+  @override
+  Future<void> resetPassword(String email) async {
+    http.Response? response = await apiService.post(
+      '/auth/request-password',
+      payload: {'email': email},
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_controller.dart]: Same comment context (transform this into a Model request)

File: lib/src/auth/auth_controller.dart [hunk 4]
```diff
@@ -1,34 +1,141 @@
...
+    http.Response? response = await apiService.post('/auth/login', payload: {
+      'email': email,
+      'password': password,
+    });
+
+    if (response != null) {
+      var jsonResponse = jsonDecode(utf8.decode(response.bodyBytes)) as Map;
+      if (response.statusCode == 200) {
+        String token = jsonResponse['data']['token'];
+        if (!JwtDecoder.isExpired(token)) {
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_controller.dart]: You should have an interceptor for this kind of things, to prevent repeating yourself in each api call, check this example from one of my projects ( https://github.com/JosePedroNobre/flutter_examples/blob/master/customer/lib/network/generic/services_base_api.dart) , im using dio but the logic should be the same

File: lib/src/auth/auth_controller.dart [hunk 5]
```diff
@@ -1,34 +1,141 @@
...
+          await storage.storeAuthToken(token);
+        }
+        apiResponse = null;
+      }
+
+      if (response.statusCode == 401) {
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_controller.dart]: This kind of things like handling generic status codes (same as the token validator) should be done on a generic approach and throw an Exception that can be caught inside the controller (and handle from there). This way we respect the DRY standard.

File: lib/src/auth/auth_controller.dart [hunk 6]
```diff
@@ -1,34 +1,141 @@
...
+    }
+
+    notifyListeners();
   }
 
-  Future<String?> readAuthToken() async {
-    return await storage.read(key: _tokenKey);
+  @override
+  Future<void> logout() async {
+    await storage.storeAuthToken('');
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_controller.dart]: storage doesn't have a .clear() method?

File: lib/src/auth/auth_controller.dart [hunk 7]
```diff
@@ -1,34 +1,141 @@
...
+    http.Response? response = await apiService.post('/auth/register', payload: {
+      'email': email,
+      'firstName': firstName,
+      'lastName': lastName,
+      'password': password,
+    });
+
+    if (response != null) {
+      var jsonResponse = jsonDecode(utf8.decode(response.bodyBytes)) as Map;
+      if (response.statusCode == 200) {
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_controller.dart]: Already discussed above

File: lib/src/auth/auth_login_form.dart [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  @override
+  State<AuthLoginForm> createState() => _AuthLoginFormState();
+}
+
+class _AuthLoginFormState extends State<AuthLoginForm> {
+  final _formKey = GlobalKey<FormState>();
+  late String _email;
+  late String _password;
+
+  void _submit(BuildContext context) async {
```
Comment (JosePedroNobre) [inline lib/src/auth/auth_login_form.dart]: Since this is a Stateful Widget , you don't need to pass the context throw parameter, you already have access to it.

File: lib/src/auth/auth_controller.dart (hunk 6 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: It has a `delete()` method so I've went ahead and used that instead.

File: lib/src/auth/auth_login_form.dart (hunk 8 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_login_form.dart]: Resolved

File: lib/src/auth/auth_controller.dart (hunk 4 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: I'm not quiet sure what you mean by having an interceptor, I'd appreciate it if you could provide me an example code snippet with more info about how you would refactor this piece of code or if you'd like, we can jump in a zoom meeting or something and just talk about it.

Comment (JosePedroNobre) [reply lib/src/auth/auth_controller.dart]: You can find that code snippet on the url I sent you, but basically an interceptor is to intercept every call and make generic logic there.Here is a url with more information ( https://medium.com/flutter-community/dio-interceptors-in-flutter-17be4214f363)

File: lib/src/auth/auth_controller.dart (hunk 6 above)
Comment (JosePedroNobre) [reply lib/src/auth/auth_controller.dart]: Nice :D

File: lib/src/auth/auth_controller.dart (hunk 3 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: Done

File: lib/src/auth/auth_controller.dart (hunk 2 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: Completed

File: lib/src/auth/auth_controller.dart (hunk 4 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: Thanks for sharing that blog post. I wasn't aware of the concept of interceptors until now so this is really interesting. I've gone ahead and replaced http with dio and added interceptors as you suggested. Please take a look at the updated code and let me know if it can be improved or optimized to work better.

File: lib/src/auth/auth_controller.dart (hunk 5 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: Completed

File: lib/src/auth/auth_controller.dart (hunk 7 above)
Comment (Andrewngabriel) [reply lib/src/auth/auth_controller.dart]: Done
